src/toxcore/dht/ping_array.rs

```diff
@@ -0,0 +1,206 @@
+/*
+    Copyright © 2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Manage ping_id.
+Generate ping_id on request packet, check ping_id on response packet.
+*/
+
+use std::io::{ErrorKind, Error};
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+
+use toxcore::crypto_core::*;
+use toxcore::dht::packed_node::*;
+
+/// Ping timeout in seconds
+pub const PING_TIMEOUT: u64 = 5;
+
+/// size of DHT ping arrays.
+pub const DHT_PING_ARRAY_SIZE: u64 = 512;
+
+/// Entry type of PingArray
+#[derive(Clone, Debug)]
+struct PingArrayEntry {
+    data: Vec<PackedNode>,
+    time: Instant,
+}
+
+/// Generate ping_id, check if ping_id is correct or timedout
+#[derive(Clone, Debug)]
+pub struct PingArray {
+    entries: HashMap<u64, PingArrayEntry>,
+    last_added: u64,
+    total_size: u64,
+    timeout: Duration,
+}
+
+impl PingArray {
+    /// Create new PingArray object
+    pub fn new(total_size: u64, timeout: Duration) -> Option<PingArray> {
+        if total_size == 0 {
+            return None;
+        }
+
+        Some(PingArray {
+            entries: HashMap::new(),
+            last_added: 0,
+            total_size,
+            timeout,
+        })
+    }
+
+    // Clear timed out entries
+    fn ping_array_clear_timedout(&mut self) {
```

Comment:
You don't need to duplicate struct name in all methods' names.

Reply:
I agree